arwen/contexts/storage.go

```diff
@@ -148,42 +156,60 @@ func (context *storageContext) GetStorageFromAddress(address []byte, key []byte)
 	// contracts themselves cannot change protected values. Values stored under
 	// protected keys must always be retrieved from the node, not from the cached
 	// StorageUpdates.
-	value, usedCache := context.getStorageFromAddressUnmetered(address, key)
+	value, usedCache, err := context.getStorageFromAddressUnmetered(address, key)
 
 	context.useExtraGasForKeyIfNeeded(key, usedCache)
 	context.useGasForValueIfNeeded(value, usedCache)
 
 	logStorage.Trace("get from address", "address", address, "key", key, "value", value)
-	return value, usedCache
+	return value, usedCache, err
 }
 
-func (context *storageContext) getStorageFromAddressUnmetered(address []byte, key []byte) ([]byte, bool) {
+func (context *storageContext) getStorageFromAddressUnmetered(address []byte, key []byte) ([]byte, bool, error) {
 	var value []byte
+	var err error
 
 	enableEpochsHandler := context.host.EnableEpochsHandler()
 	if context.isElrondReservedKey(key) && enableEpochsHandler.IsStorageAPICostOptimizationFlagEnabled() {
-		value, _, _ = context.blockChainHook.GetStorageData(address, key)
-		return value, false
+		value, err = context.readFromBlockchain(address, key)
+		return value, false, err
 	}
 
 	storageUpdates := context.GetStorageUpdates(address)
 	usedCache := true
 	if storageUpdate, ok := storageUpdates[string(key)]; ok {
 		value = storageUpdate.Data
 	} else {
-		value, _, _ = context.blockChainHook.GetStorageData(address, key)
+		value, err = context.readFromBlockchain(address, key)
+		if err != nil {
+			return nil, false, err
+		}
 		storageUpdates[string(key)] = &vmcommon.StorageUpdate{
 			Offset: key,
 			Data:   value,
 		}
 		usedCache = false
 	}
 
-	return value, usedCache
+	return value, usedCache, nil
+}
+
+func (context *storageContext) readFromBlockchain(address []byte, key []byte) ([]byte, error) {
+	value, _, err := context.blockChainHook.GetStorageData(address, key)
+	if err != nil {
+		enableEpochsHandler := context.host.EnableEpochsHandler()
+		if enableEpochsHandler.IsMaxBlockchainHookCountersFlagEnabled() {
+			return nil, err
```

Comment:
Not all errors are the same. We should only care about MaxCounters error. Other error should be ignored. Maybe ignoring them in blockchainhook level is better.

Reply:
right, will refactor as to ensure that only the counter errors passthrough